test/integration/apiserver/flowcontrol/fight_test.go

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	flowcontrol "k8s.io/api/flowcontrol/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	fcboot "k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	utilfc "k8s.io/apiserver/pkg/util/flowcontrol"
+	fqtesting "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing"
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	flowcontrolclient "k8s.io/client-go/kubernetes/typed/flowcontrol/v1beta1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/klog/v2"
+)
+
+const timeFmt = "2006-01-02T15:04:05.999"
+
+type fightTest struct {
```

Comment:
Sorry - I've spent some time trying to understand what exactly is happening in this test and so far I'm failing...

Maybe it would be beneficial to at least briefly describe what exactly you're trying to simulate...

Reply:
Attempted an explanation.

Reply:
I get (and I think I like) the controller code changes.  Reading this test, would it be simpler to use your idea of two fighting controllers without client-side ratelimiting and simply watch how many update watch events we get for flowschemas with a bound that prevents the pathological fight scenario?  That might be less mocking and slightly easier to grok.

Reply:
I am not sure I understand the suggestion.  The essence of your original test, which I kept, is to bound the _rate_ of updates.  In a scenario with fighting controllers, there will be no end to the fight --- the _number_ of updates will climb endlessly.  The _rate_ is all that I understand measuring or controlling.

Reply:
> I am not sure I understand the suggestion. The essence of your original test, which I kept, is to bound the _rate_ of updates. In a scenario with fighting controllers, there will be no end to the fight --- the _number_ of updates will climb endlessly. The _rate_ is all that I understand measuring or controlling.

It would, but you could determine a rate.  With client-side rate limiting turned off when you construct the controller, you just run it manually once without your changes to see hopefully thousands of updates.  Run it again and see hopefully dozens of updates.  Choose a threshold of something like 1000 updates in one minute.  Call it a day.

Reply:
You mean run for a fixed length of time, looking at the total number of updates over that time, right?

We would not even need to turn off client-side rate limiting; even the default rate limiter allows updates at a much higher rate than the new controller is designed to normally do.

Are you suggesting literally using one watch, or using a reflector, or an informer?  I would be inclined to use something with recovery built in.

In any case, we are not guaranteed that each update leads to a notification.  But I do not think we need a 1:1 relation there.

Do we even need to exercise the fight without the new controller reticence?  I think it would suffice to just run the regular new controllers in a fight for a fixed length of time and check that the total number of notifications is (a) not less than a bare minimum that indicates that there was a fight (perhaps something like 4 per minute) and (b) not higher than the new code is designed to allow.  That ceiling is low enough to make a real test, and also low enough to keep the CPU cost of the test down (so that we can expect that most of the updates lead to notifications, so that part (a) will not be flaky).  I assume you are suggesting to use real clocks instead of fake clocks.  The only down side I see here is that a reasonable test will take minutes.

Reply:
@MikeSpreitzer @deads2k instead of using `flowSchemaRVs map[string]string`, we might be able to reconstruct the equivalent states if we use a decorated `configQueue workqueue.RateLimitingInterface` and a decorated `flowcontrol` client (for test purpose only). This way we don't have to introduce logic in the code that is used for test purposes only.

Also, any particular reason why this test needs to be an integration test? I think we can use a unit test to simulate controller fight and then assert on our expectations.